File: T-Bank/Algorithms-and-Data-Structures-2025/12sprint-SolvingProblemMethodsPart2/contest/D-Распил_брусьев/solution.cpp
#include <iostream>
#include <vector>
#include <climits>

using namespace std;


void solution() {
    /*
    Time Complexity O(N^2)
    Memory Complexity O(N^2)
    */
    int L, N;
    scanf("%d %d", &L, &N);
    vector<int> cuts(N + 2);
    cuts[0] = 0;
    for (int i = 1; i <= N; ++i) scanf("%d", &cuts[i]);
    cuts[N + 1] = L;
    vector<vector<int>> dp(N + 2, vector<int>(N + 2, 0));
    for (int len = 2; len <= N + 1; ++len) {
        for (int left = 0; left <= N + 1 - len; ++left) {
            int right = left + len;
            dp[left][right] = INT_MAX;
            for (int k = left + 1; k < right; ++k) {
                dp[left][right] = min(dp[left][right], dp[left][k] + dp[k][right] + cuts[right] - cuts[left]);
            }
        }
    }
    printf("%d\n", dp[0][N + 1]);
}


int main() {
    solution();
}

File: T-Bank/Algorithms-and-Data-Structures-2025/12sprint-SolvingProblemMethodsPart2/contest/D-Распил_брусьев/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(N^2)
    Memory Complexity O(N^2)
    """
    L, N = map(int, sys.stdin.readline().rstrip().split())
    cuts: list[int] = [0]
    cuts.extend(map(int, sys.stdin.readline().rstrip().split()))
    cuts.append(L)
    dp: list[list[int]] = [[0] * (N + 2) for _ in range(N + 2)]
    for length in range(2, N + 2):
        for left in range(N + 1 - length + 1):
            right: int = left + length
            dp[left][right] = float('inf')
            for k in range(left + 1, right):
                dp[left][right] = min(dp[left][right], dp[left][k] + dp[k][right] + cuts[right] - cuts[left])
    sys.stdout.write('{}\n'.format(dp[0][N + 1]))


if __name__ == '__main__':
    solution()